connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -680,13 +717,13 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
             }
             sinkTaskMetricsGroup.assignedOffsets(currentOffsets);
 
-            // If we paused everything for redelivery (which is no longer relevant since we discarded the data), make
+            // If we paused everything for redelivery and all partitions for the failed deliveries have been revoked, make
             // sure anything we paused that the task didn't request to be paused *and* which we still own is resumed.
             // Also make sure our tracking of paused partitions is updated to remove any partitions we no longer own.
-            pausedForRedelivery = false;
+            pausedForRedelivery = pausedForRedelivery && !messageBatch.isEmpty();
```

Comment:
I don't know if this change is required. The way I read the current implementation, we make sure that the paused partitions contain only assigned partitions in the block below, setting the paused partitions on context. We then rely on the code block in `iteration()` to resume partitions that should not be paused.
```
            } else if (!pausedForRedelivery) {
                resumeAll();
                onResume();
            }
```
Setting this to anything other than false causes us not to resume partitions which we own that were not explicitly requested to be paused.

Reply:
> Setting this to anything other than false causes us not to resume partitions which we own that were not explicitly requested to be paused.

Yep, that's intentional. The retry loop for sink tasks works like this:
- Grab new batch of records from consumer, store in `messageBatch`
- Deliver batch to task via `SinkTask::put`
- Catch a `RetriableException` from the task
- Pause all partitions on the consumer and set `pausedForRedelivery` to `true`
- Redeliver batch (stored in the `messageBatch` field) to task via `SinkTask::put` continuously
- Once a delivery attempt is successful, clear the batch stored in `messageBatch` and, since `pausedForRedelivery` is `true`, resume all partitions on the consumer and reset `pausedForDelivery` to `false`

When consumer rebalances meant a mass-revocation and mass-reassignment of partitions, we could get away with unconditionally resetting `pausedForRedelivery` to `true` during them, since offsets for any message batches going through the retry loop would not be committed and those records would be guaranteed to be redelivered after the rebalance. However, with incremental consumer rebalances, a task might not be forced to commit offsets for some (or even all) records in its current batch (if, for example, the new assignment only included additional partitions and no revocations). In that case, we have to be careful to remain paused so that the current batch can continue to be retried.

Reply:
I'm actually wondering if there's a bug in the existing code; not sure `pausedForRedelivery` should be reset to `false` after a consumer rebalance. If the worker is in the middle of retrying a failed batch delivery, all the partitions will be paused, and after the rebalance, nothing explicitly resumes them.

During normal execution (when the connector isn't in the middle of redelivering a failed batch) this doesn't surface because partition consumption hasn't been paused anyways. But that assumption breaks down if we've paused to retry a failed batch.

I think it might be necessary to throw in a call to `resumeAll()` if `pausedForRedelivery` transitions from `true` to `false` as a result of a consumer rebalance (or, if we're lazy, we can just throw it in unconditionally at the end of each consumer rebalance if `pausedForRedelivery` is `false` since [`Consumer::resume` is a no-op for partitions that weren't already paused](https://kafka.apache.org/27/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#resume-java.util.Collection-)).

Reply:
Oooooh, one more edge case to consider--if new partitions are assigned while in the middle of retrying a failed batch, right now the worker might accidentally receive new records from those partitions, which would violate the assertion that new records can be received from the consumer if and only if the `messageBatch` field is empty.

I think we can (and should) handle this by invoking `pauseAll()` if `pausedForRedelivery` is still `true` after removing records from revoked partitions from the current batch.

Reply:
I've pushed a change that should address this gap.